sdk/include/opentelemetry/sdk/trace/batch_span_processor.h

```diff
@@ -0,0 +1,150 @@
+}
+
+namespace trace
+{   
+
+/**
+ * This is an implementation of the SpanProcessor which creates batches of finished spans and passes 
+ * the export-friendly span data representations to the configured SpanExporter.
+ */
+class BatchSpanProcessor : public SpanProcessor 
+{
+public:
+    /**
+     * Creates a batch span processor by configuring the specified exporter and other parameters
+     * as per the official, language-agnostic opentelemetry specs.
+     * 
+     * @param exporter - The backend exporter to pass the ended spans to
+     * @param max_queue_size -  The maximum buffer/queue size. After the size is reached, spans are dropped.
+     * @param schedule_delay_millis - The time interval between two consecutive exports.
+     * @param max_export_batch_size - The maximum batch size of every export. It must be smaller or equal to max_queue_size
+     */
+    explicit BatchSpanProcessor(std::unique_ptr<SpanExporter>&& exporter,
+                                const size_t max_queue_size = 2048,
+                                const std::chrono::milliseconds schedule_delay_millis = std::chrono::milliseconds(5000),
+                                const size_t max_export_batch_size = 512);
+
+    /**
+     * Requests a Recordable(Span) from the configured exporter.
+     * 
+     * @return A recordable generated by the backend exporter
+     */
+    std::unique_ptr<Recordable> MakeRecordable() noexcept override;
+
+    /**
+     * Called when a span is started.
+     * NOTE: This method is a no-op.
+     *  
+     * @param span - The span that just started
+     */
+    void OnStart(Recordable &span) noexcept override;
+
+    /**
+     * Called when a span ends.
+     * 
+     * @param span - A recordable for a span that just ended
+     */
+    void OnEnd(std::unique_ptr<Recordable> &&span) noexcept override;
+
+    /**
+     * Export all ended spans that have not been exported yet.
+     * 
+     * @param timeout - An optional timeout. A default timeout of 0 means no timeout.
+     */
+    void ForceFlush(std::chrono::microseconds timeout = std::chrono::microseconds(0)) noexcept override;
+
+    /**
+     * Shuts down the processor and does any cleanup required. Completely drains the buffer/queue of all its
+     * ended spans and passes them to the exporter. Any subsequent calls to OnStart, OnEnd, ForceFlush or
+     * Shutdown will return immediately without doing anything.
+     * 
+     * @param timeout - An optional timeout. A default timeout of 0 means no timeout.
+     */
+    void Shutdown(std::chrono::microseconds timeout = std::chrono::microseconds(0)) noexcept override;
+
+    /**
+     * Class destructor which invokes the Shutdown() method. The Shutdown() method is supposed to be invoked
+     * when the Tracer is shutdown (as per other languages), but the C++ Tracer only takes shared ownership of the processor, 
+     * and thus doesn't call Shutdown (as the processor might be shared with other Tracers).
```

Comment:
FYI @cijothomas @rajkumar-rangaraj

I think @snehilchopra is the first one doing this with the correct consideration of ref counting.